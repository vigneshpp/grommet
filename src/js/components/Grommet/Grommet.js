import React, { Component } from 'react';
import { createGlobalStyle } from 'styled-components';

import {
  ResponsiveContext,
  ThemeContext,
  ContainerTargetContext,
} from '../../contexts';

import {
  backgroundIsDark,
  deepMerge,
  getBreakpoint,
  getDeviceBreakpoint,
  normalizeColor,
} from '../../utils';
import { base as baseTheme } from '../../themes';
import { StyledGrommet } from './StyledGrommet';

const FullGlobalStyle = createGlobalStyle`
  body { margin: 0; }
`;

class Grommet extends Component {
  static displayName = 'Grommet';

  static getDerivedStateFromProps(nextProps, prevState) {
    const {
      background: backgroundProp,
      dir,
      theme = {},
      themeMode,
    } = nextProps;
    const { theme: stateTheme, themeProp, themeModeProp } = prevState;

    const nextTheme = deepMerge(baseTheme, theme);
    if (!stateTheme || theme !== themeProp || themeMode !== themeModeProp) {
      const {
        colors: { background: themeBackground },
      } = nextTheme.global;

      // get initial value for dark so we can normalize background color
      nextTheme.dark = (themeMode || theme.defaultMode) === 'dark';
      const color = normalizeColor(
        backgroundProp || themeBackground,
        nextTheme,
      );

      // After normalizing, we set nextTheme.dark once more.
      // It is necessary that we set it twice. We have to handle two cases:
      // 1. Caller passes in a color object or a color name that resolves an
      //    object. In this case, we want to set dark as line 38 shows. The
      //    second set, in line 46, is a no-op.
      // 2. Caller passes a specific color value or a color name that resolves
      //    to a specific color value. In this case, we want dark to be set
      //    based on that color, which line 46 will do.
      // The double set of nextTheme.dark allows us to handle both cases here
      // without having to duplicate color object + name + dark mode detection
      // code here that is already in normalizeColor and backgroundIsDark.
      nextTheme.dark = backgroundIsDark(color, nextTheme);
      nextTheme.baseBackground = backgroundProp || themeBackground;

      if (dir) {
        nextTheme.dir = dir;
      }
      return {
        theme: nextTheme,
        themeProp: theme,
        themeModeProp: themeMode,
      };
    }

    return null;
  }

  state = {};

  componentDidMount() {
    window.addEventListener('resize', this.onResize);
    this.onResize();
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.onResize);
  }

  onResize = () => {
    const { theme, responsive } = this.state;

    let breakpoint;
    // responsive would be undefined in the case of SSR or initial page load
    if (!responsive) {
      // In the case of SSR we'll need to use the user agent breakpoint
      breakpoint = this.deviceResponsive();
    }
    // Initial page load where both responsive and breakpoint are undefined
    if (!breakpoint) breakpoint = getBreakpoint(window.innerWidth, theme);

    if (breakpoint !== responsive) {
      this.setState({ responsive: breakpoint });
    }
  };

  deviceResponsive() {
    const { userAgent } = this.props;
    const { theme } = this.state;

    /*
     * Regexes provided for mobile and tablet detection are meant to replace
     * a full-featured specific library due to contributing a considerable size
     * into the bundle.
     *
     * User agents found https://deviceatlas.com/blog/list-of-user-agent-strings
     */
    if (userAgent) {
      if (
        /(tablet|ipad|playbook|silk)|(android(?!.*mobile))/i.test(userAgent)
      ) {
        return getDeviceBreakpoint('tablet', theme);
      }
      if (/Mobile|iPhone|Android/.test(userAgent)) {
        return getDeviceBreakpoint('phone', theme);
      }
      return getDeviceBreakpoint('computer', theme);
    }
    return undefined;
  }

  render() {
    const {
      children,
      full,
      containerTarget = typeof document === 'object'
        ? document.body
        : undefined,
      ...rest
    } = this.props;
    delete rest.theme;
    const { theme, responsive: stateResponsive } = this.state;

    // Value from state should be correct once we resize
    // On first render we try to guess otherwise set the default as a tablet
    const responsive =
      stateResponsive ||
      this.deviceResponsive() ||
      theme.global.deviceBreakpoints.tablet;

    return (
      <ThemeContext.Provider value={theme}>
        <ResponsiveContext.Provider value={responsive}>
          <ContainerTargetContext.Provider value={containerTarget}>
            <StyledGrommet full={full} {...rest}>
              {children}
            </StyledGrommet>
            {full && <FullGlobalStyle />}
          </ContainerTargetContext.Provider>
        </ResponsiveContext.Provider>
      </ThemeContext.Provider>
    );
  }
}

let GrommetDoc;
if (process.env.NODE_ENV !== 'production') {
  // eslint-disable-next-line global-require
  GrommetDoc = require('./doc').doc(Grommet);
}
const GrommetWrapper = GrommetDoc || Grommet;

export { GrommetWrapper as Grommet };
