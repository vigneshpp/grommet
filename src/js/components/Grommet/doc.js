import { describe, PropTypes } from 'react-desc';

import { backgroundDoc } from '../../utils/prop-types';
import { getAvailableAtBadge } from '../../utils/mixins';

export const doc = Grommet => {
  const DocumentedGrommet = describe(Grommet)
    .availableAt(getAvailableAtBadge('Grommet', 'Utilities'))
    .description('The top level Grommet container.')
    .usage(
      `import { Grommet } from 'grommet';
<Grommet>...</Grommet>`,
    )
    .intrinsicElement('div');

  DocumentedGrommet.propTypes = {
    background: backgroundDoc,
    dir: PropTypes.oneOf(['rtl']).description(
      'Layout direction for right to left contexts',
    ),
    full: PropTypes.bool
      .description('Whether to take the whole viewport.')
      .defaultValue(false),
    plain: PropTypes.bool
      .description(
        `Whether or not Grommet should apply a global font-family, font-size,
        and line-height.`,
      )
      .defaultValue(false),
    cssVars: PropTypes.bool
      .description('Whether to expose the css variables.')
      .defaultValue(false),
    theme: PropTypes.object.description(
      'Custom styles for Grommet app component.',
    ),
    themeMode: PropTypes.oneOf(['dark', 'light']).description(
      `Dark vs. light theme variation. Default is unspecified and left to
      theme.`,
    ),
    userAgent: PropTypes.string.description(
      `User agent used to detect the device width for setting the initial
      breakpoint.`,
    ),
    containerTarget: PropTypes.object.description(
      `The node where Drop and Layer containers are inserted. Defaults to
      document.body which is almost always the right choice. This is used
      for less common cases like rendering within an internal node (e.g.
      shadow root).`,
    ),
  };

  return DocumentedGrommet;
};

export const themeDoc = {
  'grommet.extend': {
    description: 'Any additional style for Grommet.',
    type: 'string | (props) => {}',
    defaultValue: undefined,
  },
  'global.font.face': {
    description: 'Custom font face declaration',
    type: 'string | (props) => {}',
    defaultValue: undefined,
  },
};
