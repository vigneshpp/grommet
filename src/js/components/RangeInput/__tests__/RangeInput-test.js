import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import 'jest-axe/extend-expect';
import 'regenerator-runtime/runtime';

import { axe } from 'jest-axe';
import { cleanup, render } from '@testing-library/react';
import { Grommet } from '../../Grommet';
import { RangeInput } from '..';

describe('RangeInput', () => {
  afterEach(cleanup);

  test('should have no accessibility violations', async () => {
    const { container } = render(
      <Grommet>
        <RangeInput value="50" a11yTitle="test" />
      </Grommet>,
    );

    const results = await axe(container);
    expect(results).toHaveNoViolations();
    expect(container).toMatchSnapshot();
  });

  test('renders', () => {
    const component = renderer.create(
      <Grommet>
        <RangeInput value="50" />
      </Grommet>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('track themed', () => {
    const component = renderer.create(
      <Grommet theme={{ rangeInput: { track: { color: 'brand' } } }}>
        <RangeInput value="10" />
      </Grommet>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('track themed with color and opacity', () => {
    const component = renderer.create(
      <Grommet
        theme={{ rangeInput: { track: { color: 'brand', opacity: 0.3 } } }}
      >
        <RangeInput value="10" />
      </Grommet>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('with min and max offset', () => {
    const component = renderer.create(
      <Grommet>
        <RangeInput min={10} max={20} step={1} value={15} />
      </Grommet>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
