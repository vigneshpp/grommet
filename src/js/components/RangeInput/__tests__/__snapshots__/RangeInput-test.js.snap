// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeInput renders 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  box-sizing: border-box;
  position: relative;
  -webkit-appearance: none;
  border-color: transparent;
  height: 24px;
  width: 100%;
  padding: 0px;
  cursor: pointer;
  background: transparent;
}

.c1::-moz-focus-inner {
  border: none;
}

.c1::-moz-focus-outer {
  border: none;
}

.c1::-webkit-slider-runnable-track {
  box-sizing: border-box;
  width: 100%;
  height: 4px;
  background: rgba(0,0,0,0.2);
}

.c1::-webkit-slider-thumb {
  margin-top: -10px;
  box-sizing: border-box;
  position: relative;
  border-radius: 24px;
  height: 24px;
  width: 24px;
  overflow: visible;
  background: #7D4CDB;
  -webkit-appearance: none;
  cursor: pointer;
}

.c1::-webkit-slider-thumb:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c1::-moz-range-track {
  box-sizing: border-box;
  width: 100%;
  height: 4px;
  background: rgba(0,0,0,0.2);
}

.c1::-moz-range-thumb {
  box-sizing: border-box;
  position: relative;
  border-radius: 24px;
  height: 24px;
  width: 24px;
  overflow: visible;
  background: #7D4CDB;
  -webkit-appearance: none;
  cursor: pointer;
  margin-top: 0px;
  height: 24px;
  width: 24px;
}

.c1::-ms-thumb {
  box-sizing: border-box;
  position: relative;
  border-radius: 24px;
  height: 24px;
  width: 24px;
  overflow: visible;
  background: #7D4CDB;
  -webkit-appearance: none;
  cursor: pointer;
  margin-top: 0px;
  height: 24px;
  width: 24px;
}

.c1:hover::-moz-range-thumb {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c1:hover::-ms-thumb {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c1::-ms-track {
  box-sizing: border-box;
  width: 100%;
  height: 4px;
  background: rgba(0,0,0,0.2);
  border-color: transparent;
  color: transparent;
}

.c1::-ms-fill-lower {
  background: rgba(0,0,0,0.2);
  border-color: transparent;
}

.c1::-ms-fill-upper {
  background: rgba(0,0,0,0.2);
  border-color: transparent;
}

<div
  className="c0"
>
  <input
    className="c1"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="range"
    value="50"
  />
</div>
`;

exports[`RangeInput should have no accessibility violations 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  box-sizing: border-box;
  position: relative;
  -webkit-appearance: none;
  border-color: transparent;
  height: 24px;
  width: 100%;
  padding: 0px;
  cursor: pointer;
  background: transparent;
}

.c1::-moz-focus-inner {
  border: none;
}

.c1::-moz-focus-outer {
  border: none;
}

.c1::-webkit-slider-runnable-track {
  box-sizing: border-box;
  width: 100%;
  height: 4px;
  background: rgba(0,0,0,0.2);
}

.c1::-webkit-slider-thumb {
  margin-top: -10px;
  box-sizing: border-box;
  position: relative;
  border-radius: 24px;
  height: 24px;
  width: 24px;
  overflow: visible;
  background: #7D4CDB;
  -webkit-appearance: none;
  cursor: pointer;
}

.c1::-webkit-slider-thumb:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c1::-moz-range-track {
  box-sizing: border-box;
  width: 100%;
  height: 4px;
  background: rgba(0,0,0,0.2);
}

.c1::-moz-range-thumb {
  box-sizing: border-box;
  position: relative;
  border-radius: 24px;
  height: 24px;
  width: 24px;
  overflow: visible;
  background: #7D4CDB;
  -webkit-appearance: none;
  cursor: pointer;
  margin-top: 0px;
  height: 24px;
  width: 24px;
}

.c1::-ms-thumb {
  box-sizing: border-box;
  position: relative;
  border-radius: 24px;
  height: 24px;
  width: 24px;
  overflow: visible;
  background: #7D4CDB;
  -webkit-appearance: none;
  cursor: pointer;
  margin-top: 0px;
  height: 24px;
  width: 24px;
}

.c1:hover::-moz-range-thumb {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c1:hover::-ms-thumb {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c1::-ms-track {
  box-sizing: border-box;
  width: 100%;
  height: 4px;
  background: rgba(0,0,0,0.2);
  border-color: transparent;
  color: transparent;
}

.c1::-ms-fill-lower {
  background: rgba(0,0,0,0.2);
  border-color: transparent;
}

.c1::-ms-fill-upper {
  background: rgba(0,0,0,0.2);
  border-color: transparent;
}

<div>
  <div
    class="c0"
  >
    <input
      aria-label="test"
      class="c1"
      type="range"
      value="50"
    />
  </div>
</div>
`;

exports[`RangeInput track themed 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  box-sizing: border-box;
  position: relative;
  -webkit-appearance: none;
  border-color: transparent;
  height: 24px;
  width: 100%;
  padding: 0px;
  cursor: pointer;
  background: transparent;
}

.c1::-moz-focus-inner {
  border: none;
}

.c1::-moz-focus-outer {
  border: none;
}

.c1::-webkit-slider-runnable-track {
  box-sizing: border-box;
  width: 100%;
  height: 4px;
  background: rgba(125,76,219,1);
}

.c1::-webkit-slider-thumb {
  margin-top: -10px;
  box-sizing: border-box;
  position: relative;
  border-radius: 24px;
  height: 24px;
  width: 24px;
  overflow: visible;
  background: #7D4CDB;
  -webkit-appearance: none;
  cursor: pointer;
}

.c1::-webkit-slider-thumb:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c1::-moz-range-track {
  box-sizing: border-box;
  width: 100%;
  height: 4px;
  background: rgba(125,76,219,1);
}

.c1::-moz-range-thumb {
  box-sizing: border-box;
  position: relative;
  border-radius: 24px;
  height: 24px;
  width: 24px;
  overflow: visible;
  background: #7D4CDB;
  -webkit-appearance: none;
  cursor: pointer;
  margin-top: 0px;
  height: 24px;
  width: 24px;
}

.c1::-ms-thumb {
  box-sizing: border-box;
  position: relative;
  border-radius: 24px;
  height: 24px;
  width: 24px;
  overflow: visible;
  background: #7D4CDB;
  -webkit-appearance: none;
  cursor: pointer;
  margin-top: 0px;
  height: 24px;
  width: 24px;
}

.c1:hover::-moz-range-thumb {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c1:hover::-ms-thumb {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c1::-ms-track {
  box-sizing: border-box;
  width: 100%;
  height: 4px;
  background: rgba(125,76,219,1);
  border-color: transparent;
  color: transparent;
}

.c1::-ms-fill-lower {
  background: rgba(125,76,219,1);
  border-color: transparent;
}

.c1::-ms-fill-upper {
  background: rgba(125,76,219,1);
  border-color: transparent;
}

<div
  className="c0"
>
  <input
    className="c1"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="range"
    value="10"
  />
</div>
`;

exports[`RangeInput track themed with color and opacity 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  box-sizing: border-box;
  position: relative;
  -webkit-appearance: none;
  border-color: transparent;
  height: 24px;
  width: 100%;
  padding: 0px;
  cursor: pointer;
  background: transparent;
}

.c1::-moz-focus-inner {
  border: none;
}

.c1::-moz-focus-outer {
  border: none;
}

.c1::-webkit-slider-runnable-track {
  box-sizing: border-box;
  width: 100%;
  height: 4px;
  background: rgba(125,76,219,0.3);
}

.c1::-webkit-slider-thumb {
  margin-top: -10px;
  box-sizing: border-box;
  position: relative;
  border-radius: 24px;
  height: 24px;
  width: 24px;
  overflow: visible;
  background: #7D4CDB;
  -webkit-appearance: none;
  cursor: pointer;
}

.c1::-webkit-slider-thumb:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c1::-moz-range-track {
  box-sizing: border-box;
  width: 100%;
  height: 4px;
  background: rgba(125,76,219,0.3);
}

.c1::-moz-range-thumb {
  box-sizing: border-box;
  position: relative;
  border-radius: 24px;
  height: 24px;
  width: 24px;
  overflow: visible;
  background: #7D4CDB;
  -webkit-appearance: none;
  cursor: pointer;
  margin-top: 0px;
  height: 24px;
  width: 24px;
}

.c1::-ms-thumb {
  box-sizing: border-box;
  position: relative;
  border-radius: 24px;
  height: 24px;
  width: 24px;
  overflow: visible;
  background: #7D4CDB;
  -webkit-appearance: none;
  cursor: pointer;
  margin-top: 0px;
  height: 24px;
  width: 24px;
}

.c1:hover::-moz-range-thumb {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c1:hover::-ms-thumb {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c1::-ms-track {
  box-sizing: border-box;
  width: 100%;
  height: 4px;
  background: rgba(125,76,219,0.3);
  border-color: transparent;
  color: transparent;
}

.c1::-ms-fill-lower {
  background: rgba(125,76,219,0.3);
  border-color: transparent;
}

.c1::-ms-fill-upper {
  background: rgba(125,76,219,0.3);
  border-color: transparent;
}

<div
  className="c0"
>
  <input
    className="c1"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="range"
    value="10"
  />
</div>
`;

exports[`RangeInput with min and max offset 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  box-sizing: border-box;
  position: relative;
  -webkit-appearance: none;
  border-color: transparent;
  height: 24px;
  width: 100%;
  padding: 0px;
  cursor: pointer;
  background: transparent;
}

.c1::-moz-focus-inner {
  border: none;
}

.c1::-moz-focus-outer {
  border: none;
}

.c1::-webkit-slider-runnable-track {
  box-sizing: border-box;
  width: 100%;
  height: 4px;
  background: rgba(0,0,0,0.2);
}

.c1::-webkit-slider-thumb {
  margin-top: -10px;
  box-sizing: border-box;
  position: relative;
  border-radius: 24px;
  height: 24px;
  width: 24px;
  overflow: visible;
  background: #7D4CDB;
  -webkit-appearance: none;
  cursor: pointer;
}

.c1::-webkit-slider-thumb:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c1::-moz-range-track {
  box-sizing: border-box;
  width: 100%;
  height: 4px;
  background: rgba(0,0,0,0.2);
}

.c1::-moz-range-thumb {
  box-sizing: border-box;
  position: relative;
  border-radius: 24px;
  height: 24px;
  width: 24px;
  overflow: visible;
  background: #7D4CDB;
  -webkit-appearance: none;
  cursor: pointer;
  margin-top: 0px;
  height: 24px;
  width: 24px;
}

.c1::-ms-thumb {
  box-sizing: border-box;
  position: relative;
  border-radius: 24px;
  height: 24px;
  width: 24px;
  overflow: visible;
  background: #7D4CDB;
  -webkit-appearance: none;
  cursor: pointer;
  margin-top: 0px;
  height: 24px;
  width: 24px;
}

.c1:hover::-moz-range-thumb {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c1:hover::-ms-thumb {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c1::-ms-track {
  box-sizing: border-box;
  width: 100%;
  height: 4px;
  background: rgba(0,0,0,0.2);
  border-color: transparent;
  color: transparent;
}

.c1::-ms-fill-lower {
  background: rgba(0,0,0,0.2);
  border-color: transparent;
}

.c1::-ms-fill-upper {
  background: rgba(0,0,0,0.2);
  border-color: transparent;
}

<div
  className="c0"
>
  <input
    className="c1"
    max={20}
    min={10}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    step={1}
    type="range"
    value={15}
  />
</div>
`;
